### PR TITLE
Analyzer and Code Fix for Protected Unity Message

### LINKED ISSUE #42: Recommend Unity messages be protected
#### Problem statement
Unity messages like `Awake`, `Update`, etc. do not need to be `public` for Unity to call them and are only rarely meant to be called manually.

Also, if a base class has a `private` message and a child class implements the same one, only the child one will be called without any notification or any way to call the base method, which is pretty much guaranteed to cause bugs.

#### Proposed solution
Recommend that all Unity messages be `protected`. Give a warning if they are `private`, but only a suggestion otherwise.

Inside `sealed` classes, either disable that suggestion or suppress `CS0628 // New protected member declared in sealed class`.

### PR DESCRIPTION
**Pull request by shreyalpandit**
Fixes #42

#### Short description of what this resolves:
Recommend Unity messages be protected

#### Changes proposed in this pull request:
-This adds an analyzer that makes Unity messages protected when they are not.

### CODE REVIEW

**Comment by sailro** [comment]
Perhaps something we need to figure out with @jbevain, but currently we offer in both VS and VSMac to customize the scope of our generated Unity messages: 

protected

We do not offer `Protected` so far, and the default value is `Private`, so every time a user will generate a new message (using intellisense or the wizard), we'll add this new analyzer warning. This is not a great user experience.

Perhaps we can introduce an analyzer option like `unity_style_protected_message` to dynamically enable/disable this one. This should be easy to implement on the analyzer side. (those options are already used in many analyzers like with `dotnet_style_coalesce_expression` or `dotnet_style_readonly_field`). On the IDE side when `Disable the full build` is enabled we own the execution of analyzers so we can probably set this new analyzer option to disable `UNT0021` (or only enable it if Unity message Scope is set to 'Default' ?)

File: src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		[Fact]
+		public async Task FixPrivateUnityMessage()
+		{
+			const string test = @"
+using UnityEngine;
+
+class Camera : MonoBehaviour
+{
+    private void Start()
+	{
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs:22]: I think there is a mix of tabs/spaces in this string. For all test strings, I think it is better to use spaces given we check locations.

File: src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+			var diagnostic = ExpectDiagnostic().WithLocation(6,18);
+
+			await VerifyCSharpDiagnosticAsync(test, diagnostic);
+
+			const string fixedTest = @"
+using UnityEngine;
+
+class Camera : MonoBehaviour
+{
+    protected void Start()
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs:40]: same as https://github.com/microsoft/Microsoft.Unity.Analyzers/pull/135/files#r562056310

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ProtectedUnityMessageAnalyzer.Rule.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			var declaration = root.FindNode(context.Span) as MethodDeclarationSyntax;
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:78]: root can be null and we can use pattern matching:

```suggestion
				if (!(root?.FindNode(context.Span) is MethodDeclarationSyntax declaration))
				return;
```

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
+/*--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+using System.Collections.Immutable;
+using System.Linq;
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:7]: Unused

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		}
+
+		private static async Task<Document> MakeMessageProtectedAsync(Document document, MethodDeclarationSyntax declaration, CancellationToken cancellationToken)
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+			var newDeclaration = declaration.WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.ProtectedKeyword)));
+
+			foreach (var modifier in declaration.Modifiers)
+			{
+				if (!modifier.IsKind(SyntaxKind.PublicKeyword) && !modifier.IsKind(SyntaxKind.PrivateKeyword))
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:97]: and what if i'm using an `internal` message ? Hopefully `protected internal` is a valid access modifier but not sure if we want to do that.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers

**Comment by jbevain** [comment]
Hi @shreyalpandit, thank you for the PR!

Please address the tidbits Seb pointed out. I'll meet with Seb later this week to discuss what's the best way to support this.

In the meantime, feel free to move on to the `SetPositionAndRotation` analyzer.

Thank you!

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs (hunk 5 above)
Comment (shreyalpandit) [reply src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:97]: if it's internal, would we still want to make it protected? (remove internal and add protected modifier)

Comment (sailro) [reply src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:97]: Yes let's do that to start.

File: src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+{
+	public class ProtectedUnityMessageTests : BaseCodeFixVerifierTest<ProtectedUnityMessageAnalyzer, ProtectedUnityMessageCodeFix>
+	{
+		[Fact]
+		public async Task FixPrivateUnityMessage()
+		{
+			const string test = @"
+using UnityEngine;
+
+class Camera : MonoBehaviour
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs:19]: We should detect when a class is `sealed`, in this case we should not introduce protected members given we'll add `CS0628`

File: src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+		[Fact]
+		public async Task FixPublicWithModifiersUnityMessage()
+		{
+			const string test = @"
+using UnityEngine;
+
+class Camera : MonoBehaviour
+{
+    public virtual void Awake()
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers.Tests/ProtectedUnityMessageTests.cs:61]: We have still indent/tab/space issues in other strings, I think you need to audit all test strings and fix locations accordingly.

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class ProtectedUnityMessageAnalyzer : DiagnosticAnalyzer
+	{
+		internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			id: "UNT0021",
+			title: Strings.ProtectedUnityMessageDiagnosticTitle,
+			messageFormat: Strings.ProtectedUnityMessageDiagnosticMessageFormat,
+			category: DiagnosticCategory.Correctness,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:28]: After discussing with @jbevain, we decided to mark this one with "isEnabledByDefault: false". It means that this analyzer will be `opt-in`. To enable it, users will need to add the following to their `.editorconfig` file:

```editorconfig
[*.cs]
dotnet_diagnostic.UNT0021.severity = suggestion
```

We will also update VS/VSMac tooling to offer `Protected` scope for generated messages.

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ProtectedUnityMessageAnalyzer.Rule.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+				if (!(root?.FindNode(context.Span) is MethodDeclarationSyntax declaration))
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:77]: Incorrect indent

File: src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+				return;
+			if (declaration == null)
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs:79]: Given we are using the pattern matching construct `is MethodDeclarationSyntax declaration)` we know for sure that declaration is not null

**Review by shreyalpandit — COMMENTED** [review]
rebased with main

**Comment by sailro** [comment]
Perfect. Thank you @shreyalpandit !

**Pull request merged**